file.go

```diff
@@ -0,0 +1,253 @@
+func readMetadata(pefile *pe.File, rva uint32) (string, []*Heap, error) {
+	// figure out which section contains the metadata.
+	ds := sectionByRVA(pefile, rva)
+	if ds == nil {
+		return "", nil, errors.New("metadata section is missing")
+	}
+
+	// The metadata section can be huge, we better don't call ds.Data()
+	r := ds.Open()
+
+	// seek to the virtual address specified in the COM descriptor data directory.
+	rootOffset := int64(rva - ds.VirtualAddress)
+	_, err := r.Seek(rootOffset, io.SeekStart)
+	if err != nil {
+		return "", nil, fmt.Errorf("failure to seek to the metadata root: %v", err)
+	}
+
+	read := func(data interface{}) bool {
+		err = binary.Read(r, binary.LittleEndian, data)
+		return err == nil
+	}
+	readStr := func(n int, data *string) bool {
+		buf := make([]byte, n)
+		err = binary.Read(r, binary.LittleEndian, buf)
+		if err == nil {
+			i := bytes.IndexByte(buf, 0)
+			if i == -1 {
+				i = len(buf)
+			}
+			*data = string(buf[:i])
+		}
+		return err == nil
+	}
+
+	// the Metadata header is defined in §II.24.2.1.
+	var hdr struct {
+		Signature    uint32
+		MajorVersion uint16
+		MinorVersion uint16
+		Reserved     uint32
+		Version      string
+		Flags        uint16
+	}
+	if !read(&hdr.Signature) {
+		return "", nil, fmt.Errorf("failure to read the metadata header signature: %v", err)
+	}
+	// magic signature from II.24.2.1
+	const signature = 0x424A5342
+	if hdr.Signature != signature {
+		return "", nil, fmt.Errorf("metadata header signature (%#X) must be (%#X)", hdr.Signature, signature)
+	}
+	var streamsCount uint16
+	var cstringLength uint32
+	if !read(&hdr.MajorVersion) ||
+		!read(&hdr.MinorVersion) ||
+		!read(&hdr.Reserved) ||
+		!read(&cstringLength) ||
+		!readStr(int(cstringLength), &hdr.Version) ||
+		!read(&hdr.Flags) ||
+		!read(&streamsCount) {
+		return "", nil, fmt.Errorf("failure to read the metadata header: %v", err)
+	}
+
+	readStreamNameStr := func(data *string) bool {
+		// the name of the stream as null-terminated variable length array
+		// of ASCII characters, padded to the next 4-byte boundary
+		// with \x00 characters. The name is limited to 32 characters.
+		const nameMaxLength = 32
+		const namePadding = 4
+		var nameBuf [nameMaxLength]byte
+		// Read in chunks of 4 bytes, accumulating the string
+		// into nameBuf until the first \x00 character is found.
+		var found bool
+		for j := 0; j < nameMaxLength; j += namePadding {
+			if !read(nameBuf[j : j+namePadding]) {
+				break
+			}
+			idx := bytes.IndexByte(nameBuf[:j+namePadding], 0)
+			if idx != -1 {
+				found = true
+				*data = string(nameBuf[:idx])
+				break
+			}
+		}
+		if !found {
+			err = errors.New("name not found")
+		}
+		return err == nil
+	}
+
+	// parse stream headers.
+	heaps := make([]*Heap, streamsCount)
+	for i := 0; i < int(streamsCount); i++ {
+		// the stream header is defined in §II.24.2.2.
+		var s struct {
+			Offset uint32
+			Size   uint32
+			Name   string
+		}
```

Comment:
Does a `struct` make this behave differently than three `var`s, or is it just to group them together? (I think the grouping makes sense, especially for `var hdr struct` earlier, but want to make sure my understanding is correct.)

Reply:
It's just to group them together and make it clear that we are decoding a logical group of data.